stone/frontend/ir_generator.py

```diff
@@ -468,15 +480,67 @@ def _create_annotation(self, env, item):
 
         namespace = self.api.ensure_namespace(env.namespace_name)
 
-        if item.annotation_type not in ANNOTATION_CLASS_BY_STRING:
-            raise InvalidSpec("Unknown Annotation type '%s'." %
-                              item.annotation_type, item.lineno, item.path)
+        if ((item.annotation_type_ns is None)
+                and (item.annotation_type in BUILTIN_ANNOTATION_CLASS_BY_STRING)):
+            annotation_class = BUILTIN_ANNOTATION_CLASS_BY_STRING[item.annotation_type]
+            annotation = annotation_class(item.name, namespace, item, *item.args)
+        else:
+            if item.annotation_type_ns is not None:
+                namespace.add_imported_namespace(
+                    self.api.ensure_namespace(item.annotation_type_ns),
+                    imported_data_type=True)
+
+            annotation = CustomAnnotation(item.name, namespace, item,
+                item.annotation_type, item.annotation_type_ns, item.args,
+                item.kwargs)
 
-        annotation_class = ANNOTATION_CLASS_BY_STRING[item.annotation_type]
-        annotation = annotation_class(item.name, namespace, item, *item.args)
         env[item.name] = annotation
         return annotation
 
+    def _create_annotation_type(self, env, item):
+        if item.name in env:
+            existing_dt = env[item.name]
+            raise InvalidSpec(
+                'Symbol %s already defined (%s:%d).' %
+                (quote(item.name), existing_dt._ast_node.path,
+                existing_dt._ast_node.lineno), item.lineno, item.path)
+
+        namespace = self.api.ensure_namespace(env.namespace_name)
+
+        if item.name in BUILTIN_ANNOTATION_CLASS_BY_STRING:
+            raise InvalidSpec('Cannot redefine built-in annotation type %s.' %
+                              (quote(item.name), ), item.lineno, item.path)
+
+        params = []
+        for param in item.params:
+            if param.annotations:
+                raise InvalidSpec(
+                    'Annotations cannot be applied to parameters of annotation types',
+                    param.lineno, param.path)
+            param_type = self._resolve_type(env, param.type_ref, True)
+            dt, nullable_dt = unwrap_nullable(param_type)
+
+            if isinstance(dt, Void):
+                raise InvalidSpec(
+                    'Parameters of annotation types cannot be Void',
+                    param.lineno, param.path)
+            if nullable_dt and param.has_default:
+                raise InvalidSpec(
+                    'Nullable parameters cannot have default values',
+                    param.lineno, param.path)
+            if not is_primitive_type(dt):
+                raise InvalidSpec(
+                    'Parameters of annotation types must be primitive or Nullable',
```

Comment:
I feel the error message here is a bit misleading. `Nullable` can include non-primitive types. If I understand it correctly, what we expect is either primitive or nullable primitive.